src/libdrakvuf/win-processes.c

```diff
@@ -715,6 +715,122 @@ addr_t win_get_wow_peb( drakvuf_t drakvuf, access_context_t* ctx, addr_t eproces
     return ret_peb_addr ;
 }
 
+// see https://github.com/mic101/windows/blob/master/WRK-v1.2/public/internal/base/inc/wow64tls.h#L23
+#define WOW64_TLS_CPURESERVED 1
+
+// magic offset in undocumented structure
+#define WOW64_CONTEXT_PAD 4
+
+status_t win_get_wow_context(drakvuf_t drakvuf, addr_t ethread, addr_t* wow_ctx)
+{
+    access_context_t ctx;
+    ctx.translate_mechanism = VMI_TM_PROCESS_PID;
+    ctx.pid = 4;
+    ctx.addr = ethread + drakvuf->offsets[KTHREAD_TEB];
+    addr_t teb_ptr = 0;
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &teb_ptr) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    addr_t eprocess;
+    ctx.addr = ethread + drakvuf->offsets[KTHREAD_PROCESS];
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &eprocess) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    pid_t pid;
+
+    if (drakvuf_get_process_pid(drakvuf, eprocess, &pid) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    ctx.translate_mechanism = VMI_TM_PROCESS_DTB;
+
+    if (vmi_pid_to_dtb(drakvuf->vmi, pid, &ctx.dtb) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    addr_t self_teb_ptr;
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &self_teb_ptr) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    addr_t tls_slot;
+    tls_slot = teb_ptr + drakvuf->offsets[TEB_TLS_SLOTS] + (1 * sizeof(uint64_t));
+
+    addr_t tls_slot_2;
+    ctx.addr = tls_slot;
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &tls_slot_2) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    *wow_ctx = tls_slot_2 + WOW64_CONTEXT_PAD;
+    return VMI_SUCCESS;
+}
+
+status_t win_get_stack64(drakvuf_t drakvuf, drakvuf_trap_info_t* info, addr_t* stack_ptr)
+{
+    access_context_t ctx;
+    ctx.translate_mechanism = VMI_TM_PROCESS_PID;
+    ctx.pid = 4;
+
+    addr_t ptrap_frame;
+    uint64_t rsp;
+
+    addr_t ethread = win_get_current_thread(drakvuf, info);
+    ctx.addr = ethread + drakvuf->offsets[KTHREAD_TRAPFRAME];
+
+    if (vmi_read_addr(drakvuf->vmi, &ctx, &ptrap_frame) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    ctx.addr = ptrap_frame + drakvuf->offsets[KTRAP_FRAME_RSP];
+
+    if (vmi_read_64(drakvuf->vmi, &ctx, &rsp) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    *stack_ptr = rsp;
+    return VMI_SUCCESS;
+}
+
+status_t win_get_stack32(drakvuf_t drakvuf, drakvuf_trap_info_t* info, addr_t* stack_ptr)
+{
+    addr_t ethread = win_get_current_thread(drakvuf, info);
+    addr_t wow_ctx;
+
+    if (win_get_wow_context(drakvuf, ethread, &wow_ctx) != VMI_SUCCESS)
+    {
+        return VMI_FAILURE;
+    }
+
+    access_context_t ctx;
+    ctx.translate_mechanism = VMI_TM_PROCESS_PID;
```

Comment:
In general it's always better to use `VMI_TM_PROCESS_DTB` when possible because it skips the linked-list walk of the kernel. That list is not well protected and could be tampered with.

Reply:
fixed